File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-9.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        res = False

        def dfs(node): 
            nonlocal res 

            if not node: 
                return   

            if node.val == subRoot.val: 
                res = res or self.isTheSame(node, subRoot) 

            dfs(node.left) 
            dfs(node.right) 

        dfs(root) 
        return res    



    def isTheSame(self, p, q): 
        if not p and not q: 
            return True 
        elif p and q and p.val == q.val: 
            return self.isTheSame(p.left, q.left) and self.isTheSame(p.right, q.right) 
        else: 
            return False

File: Data Structures & Algorithms/valid-binary-search-tree/submission-9.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isValidBST(root) { 

        if (!root) return false 
        return this.dfs(root, -Infinity, Infinity)
    } 

    dfs(node, left, right) {
        if (!node) return true

        if (!(left < node.val && node.val < right)) {
            return false
        } 

        let l = this.dfs(node.left, left, node.val) 
        let r = this.dfs(node.right, node.val, right) 

        return l && r
    }
}
